src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -327,15 +327,25 @@ int MPL_gpu_free_hook_register(void (*free_hook) (void *dptr))
 CUresult CUDAAPI cuMemFree(CUdeviceptr dptr)
 {
     CUresult result;
-    gpu_free_hooks_cb((void *) dptr);
-    result = sys_cuMemFree(dptr);
+    if (dptr != NULL) {
+        gpu_free_hooks_cb((void *) dptr);
+        result = sys_cuMemFree(dptr);
+    } else {
+        /* let cuMemFree deal with NULL pointer */
+        result = sys_cuMemFree(dptr);
+    }
```

Comment:
For the record, I have three suggestions:
1. without the `else` branch altogether
2. with the `else` branch explicitly because we decide we need that
3. move the `NULL` check into `gpu_free_hooks_cb` so the code here is consistent -- i.e. `gpu_free_hooks_cb` deals with the same inputs as `sys_cuMemFree`, no more no less. The bug that `gpu_free_hooks_cb` failed to deal with NULL is the bug in `gpu_free_hooks_cb`.

I am okay with any of the choice you make here. 

@minsii @raffenet One of you should make the final review.

Reply:
Sure, thanks. Now the code is doing the second suggestion. I will ask Ken to check and merge.

Reply:
I have to say of the options above, 2 is my least favorite. I would rather we have a comment to note that `gpu_free_hooks_cb` does not handle `NULL`. @minsii what do you think?

Reply:
Also, looking at the documentation `cuMemFree` makes no mention of `0` or `NULL` value as valid input. `cudaFree` explicitly does says that no operation is performed in that case.

What is the actual bug we are seeing? Is MPICH trapping a call to `cuMemFree(NULL)` from elsewhere?

Reply:
This is a trapping function to track and cache memory allocated by cuda (from user). Some application (NCCL) frees NULL pointer, which cuda internally checks and ignores (like libc free) but our hook crashes on NULL. So the fix is to add a null check here. I found the code read weird that we checked NULL then ignores the check and intentionally calls `cuMemfree` even in the NULL case. So I would like to see the code or comment explicitly explain or signal that we are aware and doing that intentionally.

Reply:
So here is what I think after reading through the discussion and looking at the cuda docs. We should never check for `NULL` in our `cuMemFree` trap. It is required to be a valid pointer value. If we check for `NULL`, it will hide bugs in the user program that may cause crashes with other (non-MPICH) builds.

The real issue is then in our trap for `cudaFree`, which may contain `NULL` in `dptr`. The 2 options I like are:
1. allow `NULL` in `gpu_free_hooks_cb`
2. check for `NULL` in `cudaFree` trap and return early, skipping hooks and `sys_cudaFree`

Reply:
Both are my suggestions 3 and 1. Both are fine with me.

Reply:
I also would accept skipping hooks in case of `NULL` and still calling `cudaFree` (eliminate else suggestion from above). At that point it really doesn't matter, so I don't feel that strongly.

Reply:
Sorry for joining the discussion late. I think it is better to allow NULL for `cudaFree`, as it is a valid input. The real bug here seems to be in `gpu_free_hooks_cb`. Rather than changing `cudaFree`, I'd recommend we fix `gpu_free_hooks_cb` (e.g., early return if `!dptr`). My suggestion `==option 1` in @raffenet 's suggestion and `==option 3` in @hzhou's suggestion.